### PR TITLE
<common> fix formatter buffer out-of-bounds

### PR DESCRIPTION
**Pull request by VVoidV**
The length of buf in Formatter is 1024. If the output was truncated due to size limit, then the return value is the number of characters (excluding the terminating null byte) which would have been written to the final string if enough space had been available. So XMLFormatter::dump_format_va will create a string_view that out of bounds of buf.
https://tracker.ceph.com/issues/68658

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by VVoidV** [comment]
@cbodley  Can you take a look at this? The CompleteMultipartUpload response dumps the Location of the object, which is base_url/bucket/obj_key. This element can exceed a 1024-length buffer. When I perform a multipart upload with a key length of 1011, I found the Location element is incomplete and trailing with garbage content.

https://github.com/ceph/ceph/blob/f2a7ed4952c8591a5c407c1e58850a7d41776825/src/rgw/rgw_rest_s3.cc#L4200-L4235

## example
```
<?xml version="1.0" encoding="UTF-8" ?>
<CompleteMultipartUploadResult xmlns="http://s3.amazonaws.com/doc/2006-03-01/">
   <Location>http: //127.0.0.1:8000/test-lbr/kvscycebdquwbwfqwpgjagcqyqojugkowjjidfnpzaqogjybyzlnevdlluygplgpatdevtidlrokcigpikrwzpkaejjyjsbuzqtagnpeprstohgforymfyoulmzptzpafxcumnjkybgtmcrknazxmxpnhnlpebatrzxgzjcnpdlqppwxeadzrxgzxgbcbxjlnubsptkytlunftenunlsnlnwsjtsuiuvhnmzfnakbxcmpvqanlispceshsnqlnmzjzhpicmwmgacetxzbcipkkzlaefchkoveutdgruuolllgwvqahtvpovtaxxeeroszjxnvmjceueuhkpphxacdahcvdgfbnmcurcirhlcvwfpwsmtfcnxqtwfvzzquolmsgsyzvnqbkepjbomnrjcbppxtzlicifbaxpzxbpvhpnxvpgjfvyajxkutqjyilzqmywrbtetewpheuibrbuohnpbvjmlubisztjapuolhaszgzxvqgbkmemhtmnnnvhhpeyughqoeeeuruqodzpmtgdiattxzyatjrkcnndwzogigtmjvrdosdymymbptqdivdonhqrleeukvzywowxpgbmldqbtrlrszxjwuhbkrlyxarwgsjnbvjkiptdxagwngaxygzbgygzljrbiwpgjdtmrkjptildtyjlqkbieencynalquukrhczruvaxrhhaowuagxvjriecqupghpjpddokkwdojdpllydajswxuhnrerjvdqwfviwnrpylzkywpcdxlflsrcdwmcjyhkvrjaoiwmogtykvrhvzwyvsrdyvuctybyggqtwadbngiqlobptopeujdnwcsropjpoomgnfgtenttfspljcdyxnwbytzxkqvzunjmkcvtcnqshgbecpoggmhhjnmrqhjzvqhbsfyvhavwkbgckywhlanyqvnrkyrgwohrwxvapagljeijsyvhdadcrpnrtkyknhaxjlpmmqszth&#x00;&#x18;\xd8\x97\xb3&#x04;V&#x00;&#x00;&#x00;j\x81{\xe1V&#x1e;&#x08;@\x8c</Location>
   <Bucket>test-lbr</Bucket>
   <Key>kvscycebdquwbwfqwpgjagcqyqojugkowjjidfnpzaqogjybyzlnevdlluygplgpatdevtidlrokcigpikrwzpkaejjyjsbuzqtagnpeprstohgforymfyoulmzptzpafxcumnjkybgtmcrknazxmxpnhnlpebatrzxgzjcnpdlqppwxeadzrxgzxgbcbxjlnubsptkytlunftenunlsnlnwsjtsuiuvhnmzfnakbxcmpvqanlispceshsnqlnmzjzhpicmwmgacetxzbcipkkzlaefchkoveutdgruuolllgwvqahtvpovtaxxeeroszjxnvmjceueuhkpphxacdahcvdgfbnmcurcirhlcvwfpwsmtfcnxqtwfvzzquolmsgsyzvnqbkepjbomnrjcbppxtzlicifbaxpzxbpvhpnxvpgjfvyajxkutqjyilzqmywrbtetewpheuibrbuohnpbvjmlubisztjapuolhaszgzxvqgbkmemhtmnnnvhhpeyughqoeeeuruqodzpmtgdiattxzyatjrkcnndwzogigtmjvrdosdymymbptqdivdonhqrleeukvzywowxpgbmldqbtrlrszxjwuhbkrlyxarwgsjnbvjkiptdxagwngaxygzbgygzljrbiwpgjdtmrkjptildtyjlqkbieencynalquukrhczruvaxrhhaowuagxvjriecqupghpjpddokkwdojdpllydajswxuhnrerjvdqwfviwnrpylzkywpcdxlflsrcdwmcjyhkvrjaoiwmogtykvrhvzwyvsrdyvuctybyggqtwadbngiqlobptopeujdnwcsropjpoomgnfgtenttfspljcdyxnwbytzxkqvzunjmkcvtcnqshgbecpoggmhhjnmrqhjzvqhbsfyvhavwkbgckywhlanyqvnrkyrgwohrwxvapagljeijsyvhdadcrpnrtkyknhaxjlpmmqszthtmmnhshzxvxuftixctb</Key>
   <ETag>0cdaffc8e69cab00470bf84bc31f305b-4</ETag>
</CompleteMultipartUploadResult>
```

File: src/common/Formatter.cc [hunk 1]
```diff
@@ -366,9 +366,24 @@ std::ostream& JSONFormatter::dump_stream(std::string_view name)
...
+  int len = vsnprintf(final_buf, buf_size, fmt, ap_copy);
+  va_end(ap_copy);
+
+  std::unique_ptr<char[]> large_buf;
+  if (len >= buf_size) {
+    // output was truncated, allocate a buffer large enough
+    buf_size = len + 1;
+    large_buf = std::make_unique<char[]>(buf_size);
+    vsnprintf(large_buf.get(), buf_size, fmt, ap);
+    final_buf = large_buf.get();
```
Comment (cbodley) [inline src/common/Formatter.cc:383]: i really like boost's [small_vector](https://www.boost.org/doc/libs/1_85_0/doc/html/boost/container/small_vector.html#id-1.3.9.14.37.8.4) for things like this. it has the same stack vs. heap behavior, but saves you from having to manage them separately. what do you think?
```c++
  auto buf = boost::container::small_vector<char, LARGE_SIZE>{
      LARGE_SIZE, boost::container::default_init};

  va_list ap_copy;
  va_copy(ap_copy, ap);
  int len = vsnprintf(buf.data(), buf.size(), fmt, ap_copy);
  va_end(ap_copy);

  if (std::cmp_greater_equal(len, buf.size())) {
    // output was truncated, allocate a buffer large enough
    buf.resize(len + 1, boost::container::default_init);
    vsnprintf(buf.data(), buf.size(), fmt, ap);
  }
```

**Review by cbodley — COMMENTED** [review]
the use of `va_copy()` and `vsnprintf()` looks correct to me 👍

File: src/common/Formatter.cc (hunk 1 above)
Comment (VVoidV) [reply src/common/Formatter.cc]: Thanks for your kindness. I've applied the suggestion and found another HTMLFormatter that needs fixing. The unittests have also been added. @cbodley

**Review by cbodley — APPROVED** [review]

**Comment by cbodley** [comment]
https://jenkins.ceph.com/job/ceph-pull-requests/145645/
> The Jenkins agent crashed or was restarted

**Comment by cbodley** [comment]
jenkins test make check

**Comment by cbodley** [comment]
https://jenkins.ceph.com/job/ceph-api/83929/
> ERROR: setUpClass (tasks.mgr.dashboard.test_health.HealthTest)

tracked in https://tracker.ceph.com/issues/47612

**Comment by cbodley** [comment]
jenkins test api

**Comment by cbodley** [comment]
jenkins test windows

**Comment by cbodley** [comment]
jenkins test api

**Comment by cbodley** [comment]
passed qa in https://pulpito.ceph.com/cbodley-2024-12-13_15:29:17-rgw-wip-cbodley2-testing-distro-default-smithi/

**Comment by cbodley** [comment]
tyvm @VVoidV, sorry for the delay

**Pull request merged**